File: android/src/main/java/com/mycompany/myrubikscube/android/AndroidLauncher.java
package com.mycompany.myrubikscube.android;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;

import com.badlogic.gdx.backends.android.AndroidApplication;
import com.badlogic.gdx.backends.android.AndroidApplicationConfiguration;
import com.mycompany.myrubikscube.CubeApp;
import com.mycompany.myrubikscube.PlatformBridge;

public class AndroidLauncher extends AndroidApplication implements PlatformBridge {

    private static final String TAG = "AndroidLauncher";
    private static final String EXTRA_CUBE_STRING = "CUBE_STRING";
    private CubeApp cubeApp;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        AndroidApplicationConfiguration config = new AndroidApplicationConfiguration();

        // Initialize CubeApp
        cubeApp = new CubeApp(this);
        initialize(cubeApp, config);

        // Delay processing the cube string until CubeApp is initialized
        new Handler().postDelayed(() -> {
            Intent intent = getIntent();
            if (intent != null && intent.hasExtra(EXTRA_CUBE_STRING)) {
                String cubeString = intent.getStringExtra(EXTRA_CUBE_STRING);
                Log.d(TAG, "Received Cube String: " + cubeString);
                cubeApp.onCubeScanned(cubeString); // This ensures cube is initialized before calling
            }
        }, 500); // Give time for CubeApp to fully initialize
    }


    @Override
    public void startColorInputActivity() {
        Intent intent = new Intent(this, ColorInputActivity.class);
        startActivity(intent);
    }


    /**
     * Receives the cube string from ColorInputActivity.
     */
    @Override
    public void sendCubeString(String cubeString) {
        cubeApp.onCubeScanned(cubeString);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (intent != null && intent.hasExtra(EXTRA_CUBE_STRING)) {
            String cubeString = intent.getStringExtra(EXTRA_CUBE_STRING);
            Log.d(TAG, "Received Cube String via onNewIntent: " + cubeString);
            cubeApp.onCubeScanned(cubeString);
        }
    }
}

File: android/src/main/java/com/mycompany/myrubikscube/android/AndroidLauncher.kt
package com.mycompany.myrubikscube.android

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import com.badlogic.gdx.backends.android.AndroidApplication
import com.badlogic.gdx.backends.android.AndroidApplicationConfiguration
import com.mycompany.myrubikscube.CubeApp
import com.mycompany.myrubikscube.PlatformBridge

class AndroidLauncher : AndroidApplication(), PlatformBridge {

    companion object {
        const val EXTRA_CUBE_STRING = "CUBE_STRING"
        const val REQUEST_CODE_SCAN = 1
    }

    private val TAG = "AndroidLauncher"
    private lateinit var cubeApp: CubeApp

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val config = AndroidApplicationConfiguration()

        val cubeString = intent.getStringExtra(EXTRA_CUBE_STRING)

        cubeApp = CubeApp(this, cubeString)
        initialize(cubeApp, config)
    }

    override fun startColorInputActivity() {
        val intent = Intent(this, ColorInputActivity::class.java)
        startActivityForResult(intent, REQUEST_CODE_SCAN)
    }

    /**
     * Receives the cube string from ColorInputActivity.
     */
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_SCAN && resultCode == RESULT_OK && data != null) {
            if (data.hasExtra(EXTRA_CUBE_STRING)) {
                val cubeString = data.getStringExtra(EXTRA_CUBE_STRING)
                Log.d(TAG, "Received Cube String via onActivityResult: $cubeString")
                cubeApp.gameScreen.onCubeScanned(cubeString ?: "")
            }
        }
    }

    override fun sendCubeString(cubeString: String?) {
        cubeApp.gameScreen.onCubeScanned(cubeString ?: "")
    }

    override fun showMessage(message: String) {
        runOnUiThread {
            Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCubeSolved() {
        runOnUiThread {
            Toast.makeText(this, "Cube Solved!", Toast.LENGTH_SHORT).show()
        }
    }

    override fun handleCubeMessage(msg: String) {
        showMessage(msg)
    }

    override fun handleCubeSolved() {
        onCubeSolved()
    }

    override fun onAlgorithmCompleted() {
        cubeApp.gameScreen.onAlgorithmCompleted()
    }

    override fun handleMainMenu() {
        finish()
    }
}
